runtime/j9vm/valhallavmi.cpp

```diff
@@ -119,5 +138,99 @@ JVM_NewNullRestrictedArray(JNIEnv *env, jclass componentType, jint length)
 	vmFuncs->internalExitVMToJNI(currentThread);
 	return arrayRef;
 }
+
+JNIEXPORT jarray JNICALL
+JVM_NewNullRestrictedAtomicArray(JNIEnv *env, jclass componentType, jint length)
+{
+	J9VMThread *currentThread = (J9VMThread *)env;
+	J9JavaVM *vm = currentThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = currentThread->javaVM->internalVMFunctions;
+	J9Class *ramClass = NULL;
+	j9object_t newArray = NULL;
+	jarray arrayRef = NULL;
+
+	vmFuncs->internalEnterVMFromJNI(currentThread);
+	ramClass = J9VMJAVALANGCLASS_VMREF(currentThread, J9_JNI_UNWRAP_REFERENCE(componentType));
+
+	if (length < 0) {
+		vmFuncs->setCurrentException(currentThread, J9VMCONSTANTPOOL_JAVALANGILLEGALARGUMENTEXCEPTION, NULL);
+		goto done;
+	}
+
+	if (!(J9_IS_J9CLASS_VALUETYPE(ramClass) && J9_IS_J9CLASS_ALLOW_DEFAULT_VALUE(ramClass))) {
+		vmFuncs->setCurrentException(currentThread, J9VMCONSTANTPOOL_JAVALANGILLEGALARGUMENTEXCEPTION, NULL);
+		goto done;
+	}
+
+	if (NULL == J9CLASS_GET_NULLRESTRICTED_ARRAY(ramClass)) {
+		J9ROMArrayClass *arrayOfObjectsROMClass = (J9ROMArrayClass *)J9ROMIMAGEHEADER_FIRSTCLASS(vm->arrayROMClasses);
+		vmFuncs->internalCreateArrayClassWithOptions(
+			currentThread, arrayOfObjectsROMClass, ramClass, J9_FINDCLASS_FLAG_CLASS_OPTION_NULL_RESTRICTED_ARRAY);
+		if (NULL != currentThread->currentException) {
+			goto done;
+		}
+		ramClass = VM_VMHelpers::currentClass(ramClass);
+	}
+
+	newArray = vm->memoryManagerFunctions->J9AllocateIndexableObject(
+			currentThread, J9CLASS_GET_NULLRESTRICTED_ARRAY(ramClass), length, J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
+
+	if (NULL == newArray) {
+		vmFuncs->setHeapOutOfMemoryError(currentThread);
+		goto done;
+	}
+
+	arrayRef = (jarray)vmFuncs->j9jni_createLocalRef(env, newArray);
+done:
+	vmFuncs->internalExitVMToJNI(currentThread);
+	return arrayRef;
+}
+
+JNIEXPORT jarray JNICALL
+JVM_NewNullableAtomicArray(JNIEnv *env, jclass componentType, jint length)
```

Comment:
given this is a common pattern for `JVM_NewNullRestrictedArray`, `JVM_NewNullRestrictedAtomicArray` and `JVM_NewNullableAtomicArray` you can make a common helper that takes a UDATA to determine array type.

```
#define NULLRESTRICTEDARRAY 0
...
```